File: bench/atlas_bench/ids.py
"""Identifier computation (SPEC §2).

Every id in the Atlas is derived, never typed by hand:

* ``config_id`` — ``sha256(canonical)[:16]``
* ``cell_id``   — ``sha256("model|quant|hardware|count|engine|engine_minor")[:12]``
* ``run_id``    — ``<config_id>--<workload_id>--<sha256(login|started_at)[:6]>``
* result path   — ``results/<engine>/<model>/<hardware>/<run_id>.json``
"""

from __future__ import annotations

import hashlib
import re
from pathlib import PurePosixPath
from typing import Any

from .canonical import CanonicalInput, canonicalize

__all__ = [
    "ID_RE",
    "cell_id",
    "config_id",
    "engine_minor",
    "is_valid_id",
    "result_path",
    "run_id",
    "run_suffix",
    "sha256_hex",
    "slugify_id",
]

#: Registry ids are lowercase kebab-case (SPEC §2).
ID_RE = re.compile(r"^[a-z0-9][a-z0-9.-]*$")


def sha256_hex(text: str) -> str:
    """Hex sha256 of the UTF-8 encoding of ``text``."""
    return hashlib.sha256(text.encode("utf-8")).hexdigest()


def config_id(inp: CanonicalInput) -> str:
    """16 hex chars of ``sha256`` over the canonical config string."""
    return sha256_hex(canonicalize(inp))[:16]


def config_id_from_canonical(canonical: str) -> str:
    """16 hex chars of ``sha256`` over an already canonicalized string."""
    return sha256_hex(canonical)[:16]


_MINOR_RE = re.compile(r"^v?(\d+)\.(\d+)")


def engine_minor(version: str) -> str:
    """First two components of a version.

    ``0.27.1 → 0.27``, ``v1.2.3 → 1.2``, ``0.26.1.dev0+g568afb3a1 → 0.26``. Version schemes
    that are not dotted numbers (llama.cpp's ``b7000``, LM Studio's dates) have no minor to
    speak of and are returned lowercased and trimmed, so each build is its own square.
    """
    cleaned = str(version).strip().lower()
    match = _MINOR_RE.match(cleaned)
    return f"{match.group(1)}.{match.group(2)}" if match else cleaned


def cell_id(
    *,
    model_id: str,
    quant_id: str,
    hardware_id: str,
    hw_count: int,
    engine_id: str,
    engine_version: str,
) -> str:
    """12 hex chars identifying one square of the coverage map."""
    key = "|".join(
        [
            model_id,
            quant_id,
            hardware_id,
            str(int(hw_count)),
            engine_id,
            engine_minor(engine_version),
        ]
    )
    return sha256_hex(key)[:12]


def run_suffix(github_login: str, started_at: str) -> str:
    """6 hex chars distinguishing two runs of the same config+workload."""
    return sha256_hex(f"{github_login}|{started_at}")[:6]


def run_id(*, cfg_id: str, workload_id: str, github_login: str, started_at: str) -> str:
    """``<config_id>--<workload_id>--<suffix>``."""
    return f"{cfg_id}--{workload_id}--{run_suffix(github_login, started_at)}"


def result_path(*, engine_id: str, model_id: str, hardware_id: str, rid: str) -> PurePosixPath:
    """Repo-relative path a result file must live at."""
    return PurePosixPath("results") / engine_id / model_id / hardware_id / f"{rid}.json"


def is_valid_id(value: Any) -> bool:
    """True when ``value`` is a well-formed registry id."""
    return isinstance(value, str) and bool(ID_RE.match(value))


def slugify_id(text: str) -> str:
    """Best-effort conversion of a free-form name into a registry-shaped id."""
    slug = re.sub(r"[^a-z0-9.]+", "-", text.strip().lower())
    slug = re.sub(r"-{2,}", "-", slug).strip("-.")
    return slug or "unknown"

File: packages/core/src/ids.ts
import { sha256Hex } from './hash.js';

/**
 * The computed identifiers — SPEC §2.
 *
 * Every one of these is derived, never typed by a human, and recomputed by the validator.
 * If a contributor edits a number by hand, the ids stop matching and CI fails.
 */

export interface CellIdInput {
  model_id: string;
  quant_id: string;
  hardware_id: string;
  hw_count: number;
  engine_id: string;
  /** First two version components, e.g. `0.27`. Use `engineMinor()` to derive it. */
  engine_minor: string;
}

/**
 * `sha256("model|quant|hardware|count|engine|minor")[:12]` — one square of the coverage map.
 *
 * The engine *minor* rather than the full version is deliberate: patch releases of an engine
 * are the same square, minors are not, because 0.26 → 0.27 routinely moves numbers by double
 * digit percentages.
 */
export function cellId(input: CellIdInput): string {
  const parts = [
    input.model_id,
    input.quant_id,
    input.hardware_id,
    String(input.hw_count),
    input.engine_id,
    input.engine_minor,
  ];
  return sha256Hex(parts.join('|')).slice(0, 12);
}

/**
 * First two components of a version.
 *
 * `0.27.1 → 0.27`, `v1.2.3 → 1.2`, `0.26.1.dev0+g568afb3a1 → 0.26`. Version schemes that are
 * not dotted numbers (llama.cpp's `b7000`, LM Studio's dates) have no minor to speak of and
 * are returned lowercased and trimmed, so each build is its own square.
 */
export function engineMinor(version: string): string {
  const v = version.trim().toLowerCase();
  const m = /^v?(\d+)\.(\d+)/.exec(v);
  return m ? `${m[1]}.${m[2]}` : v;
}

/** `<config_id>--<workload_id>--<sha256(login|started_at)[:6]>` */
export function runId(
  configId: string,
  workloadId: string,
  githubLogin: string,
  startedAt: string,
): string {
  const who = sha256Hex(`${githubLogin}|${startedAt}`).slice(0, 6);
  return `${configId}--${workloadId}--${who}`;
}

export interface ParsedRunId {
  configId: string;
  workloadId: string;
  contributorHash: string;
}

/**
 * Inverse of `runId` as far as it goes. Returns null for anything that is not a run id.
 * Workload ids may contain single dashes but never `--`, which is what makes this unambiguous.
 */
export function parseRunId(id: string): ParsedRunId | null {
  const parts = id.split('--');
  if (parts.length !== 3) return null;
  const [configId, workloadId, contributorHash] = parts as [string, string, string];
  if (!/^[0-9a-f]{16}$/.test(configId)) return null;
  if (!/^[0-9a-f]{6}$/.test(contributorHash)) return null;
  if (!/^[a-z0-9][a-z0-9.-]*$/.test(workloadId)) return null;
  return { configId, workloadId, contributorHash };
}

/** `results/<engine>/<model>/<hardware>/<run_id>.json` — the only place a result may live. */
export function resultPath(
  engineId: string,
  modelId: string,
  hardwareId: string,
  id: string,
): string {
  return `results/${engineId}/${modelId}/${hardwareId}/${id}.json`;
}

/** Directory a packet tells an agent to write into. */
export function resultDir(engineId: string, modelId: string, hardwareId: string): string {
  return `results/${engineId}/${modelId}/${hardwareId}`;
}
